File: dart/transaction_system/dart_application_1/bin/inventory_service.dart
import 'package:uuid/uuid.dart';
// lib/services/inventory_service.dart

import 'models/inventory_item.dart';
import 'inventory_repository.dart';
import 'audit_logger.dart';

class InventoryService {
  final InventoryRepository _repo;
  final AuditLogger _auditLogger;
  final Uuid _uuid;

  InventoryService({
    required InventoryRepository repo,
    required AuditLogger auditLogger,
  })  : _repo = repo,
        _auditLogger = auditLogger,
        _uuid = const Uuid();

  List<InventoryItem> getAllItems() => _repo.getAll();

  InventoryItem? getById(String id) => _repo.findById(id);

  InventoryItem? getBySku(String sku) => _repo.findBySku(sku);

  List<InventoryItem> getLowStock({int threshold = 5}) =>
      _repo.getLowStock(threshold);

  /// Creates a new item. Returns null and a message if SKU already exists.
  (InventoryItem?, String?) createItem({
    required String sku,
    required String name,
    required double unitCost,
    int initialQuantity = 0,
    String? description,
    String? createdBy,
  }) {
    if (_repo.skuExists(sku)) {
      return (null, 'SKU already exists: $sku');
    }

    final item = InventoryItem(
      id: _uuid.v4(),
      sku: sku,
      name: name,
      quantity: initialQuantity,
      unitCost: unitCost,
      lastUpdated: DateTime.now(),
      description: description,
    );

    _repo.save(item);
    _auditLogger.logItemCreated(item, userId: createdBy);
    return (item, null);
  }

  /// Updates item metadata (not quantity — use TransactionService for that).
  (InventoryItem?, String?) updateItem(
    String id, {
    String? name,
    double? unitCost,
    String? description,
    String? updatedBy,
  }) {
    final existing = _repo.findById(id);
    if (existing == null) return (null, 'Item not found: $id');

    final updated = existing.copyWith(
      name: name,
      unitCost: unitCost,
      description: description,
      lastUpdated: DateTime.now(),
    );

    _repo.save(updated);
    _auditLogger.logItemUpdated(existing, updated, userId: updatedBy);
    return (updated, null);
  }

  /// Soft-archives an item. Returns false if not found.
  (bool, String?) deleteItem(String id, {String? deletedBy}) {
    final item = _repo.findById(id);
    if (item == null) return (false, 'Item not found: $id');

    _auditLogger.logItemDeleted(item, userId: deletedBy);
    _repo.delete(id);
    return (true, null);
  }
}

File: python system/inventory_system/services/inventory_service.py
import uuid
from typing import Optional, List, Tuple
from datetime import datetime
from models.inventory_item import InventoryItem
from repositories.inventory_repository import InventoryRepository
from services.audit_logger import AuditLogger


class InventoryService:
    def __init__(self, repo: InventoryRepository, audit_logger: AuditLogger):
        self._repo = repo
        self._audit_logger = audit_logger

    def get_all_items(self) -> List[InventoryItem]:
        return self._repo.get_all()

    def get_by_id(self, id: str) -> Optional[InventoryItem]:
        return self._repo.find_by_id(id)

    def get_by_sku(self, sku: str) -> Optional[InventoryItem]:
        return self._repo.find_by_sku(sku)

    def get_low_stock(self, threshold: int = 5) -> List[InventoryItem]:
        return self._repo.get_low_stock(threshold)

    def create_item(
        self,
        sku: str,
        name: str,
        unit_cost: float,
        initial_quantity: int = 0,
        description: Optional[str] = None,
        created_by: Optional[str] = None,
    ) -> Tuple[Optional[InventoryItem], Optional[str]]:
        if self._repo.sku_exists(sku):
            return None, f"SKU already exists: {sku}"

        item = InventoryItem(
            id=str(uuid.uuid4()),
            sku=sku,
            name=name,
            quantity=initial_quantity,
            unit_cost=unit_cost,
            last_updated=datetime.now(),
            description=description,
        )

        self._repo.save(item)
        self._audit_logger.log_item_created(item, user_id=created_by)
        return item, None

    def update_item(
        self,
        id: str,
        name: Optional[str] = None,
        unit_cost: Optional[float] = None,
        description: Optional[str] = None,
        updated_by: Optional[str] = None,
    ) -> Tuple[Optional[InventoryItem], Optional[str]]:
        existing = self._repo.find_by_id(id)
        if existing is None:
            return None, f"Item not found: {id}"

        updated = existing.copy_with(
            name=name,
            unit_cost=unit_cost,
            description=description,
            last_updated=datetime.now(),
        )

        self._repo.save(updated)
        self._audit_logger.log_item_updated(existing, updated, user_id=updated_by)
        return updated, None

    def delete_item(self, id: str, deleted_by: Optional[str] = None) -> Tuple[bool, Optional[str]]:
        item = self._repo.find_by_id(id)
        if item is None:
            return False, f"Item not found: {id}"

        self._audit_logger.log_item_deleted(item, user_id=deleted_by)
        self._repo.delete(id)
        return True, None